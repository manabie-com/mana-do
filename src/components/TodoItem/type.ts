import React from 'react'
import type {Todo} from '../../models/todo'

export type TodoItemProps = {
  todo: Todo
  onChangeStatus: (event: React.ChangeEvent<HTMLInputElement>, todoId: string) => void
  onDelete: (event: React.MouseEvent<HTMLButtonElement>, todoId: string) => void
}