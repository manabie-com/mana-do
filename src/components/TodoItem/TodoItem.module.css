.root {
  display: flex;
  align-items: center;
  justify-content: space-between;
  min-height: 36px;
  margin-bottom: 6px;
}
.root:last-child {
  margin-bottom: 0;
}

.root > p {
  flex: 1 1;
  text-align: left;
  margin: 0;
  margin-left: 8px;
  cursor: pointer;
}
.todoDelete {
  outline: none;
  border: none;
  width: 32px;
  height: 32px;
  min-width: auto;
  min-height: auto;
  box-shadow: none;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
}