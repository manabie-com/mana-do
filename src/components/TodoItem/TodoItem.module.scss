.deleteButton {
  display: none;
  position: absolute;
  top: 0;
  right: 12px;
  bottom: 0;
  width: 30px;
  height: 30px;
  top: 50%;
  transform: translate(0, -15px);
  font-size: 35px;
  color: #cc9a9a;
  transition: color 0.2s ease-out;
  background: none;
  outline: none;
  border: none;
  padding: 0;
  cursor: pointer;
  &:hover {
    color: #af5b5e;
  }
  &::after {
    content: "Ã—";
  }
}

.item {
  display: flex;
  align-items: center;
  position: relative;
  font-size: 24px;
  border-bottom: 1px solid #ededed;
  justify-content: space-between;
  width: 100%;
  padding: 4px 20px;
  &:hover {
    background-color: #fafafa;
    .deleteButton {
      display: block;
    }
  }
  label {
    width: 100%;
    text-align: left;
    word-break: break-all;
    padding: 15px 0;
    display: block;
    line-height: 1.2;
    transition: color 0.4s;
    font-weight: 400;
    color: #4d4d4d;
    max-width: 100%;
    margin-bottom: 5px;
    display: flex;
    align-items: center;
    gap: 10px;
    margin: 0;
    &.completed {
      color: #cdcdcd;
      text-decoration: line-through;
    }
  }

  input[type="checkbox"] {
    width: 24px;
    height: 24px;
    box-shadow: none;
    border: none;
    outline: none;
    flex: 0 0 24px;
    cursor: pointer;
  }

  input[type="checkbox"]:focus {
    box-shadow: none;
    border: none;
    outline: none;
  }
}

.editInput {
  padding: 12px;
  border: 2px solid #ededed;
  background: rgba(0, 0, 0, 0.003);
  box-shadow: inset 0 -2px 1px rgba(0, 0, 0, 0.03);
  position: relative;
  margin: 0;
  width: 100%;
  font-size: 24px;
  line-height: 1.4em;
  font-style: italic;
  outline: none;
}
