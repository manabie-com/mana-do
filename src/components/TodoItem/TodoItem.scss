@import "../../assets/styles/main.scss";

.Todo__item {
  padding: 0.5rem 1rem;
  height: 3rem;
  display: flex;
  align-items: center;
  background-color: #ffffff;
  margin: 0.5rem 0;
  border-radius: 1rem;
  box-shadow: rgba(149, 157, 165, 0.2) 0px 0.5rem 1.5rem;
  font-size: 14px;

  &_checkbox {
    position: relative;
    display: flex;
    margin-left: -1.5rem;
  }

  &_input {
    width: auto;
    max-width: 100%;
    position: relative;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    margin-left: -0.5rem;
    margin-right: 1rem;

    &--fullwidth {
      width: 100%;
    }

    &--modify {
      width: 100%;
      padding: 0;
    }
  }
  &_button {
    &--delete {
      outline: none;
      border: none;
      width: 2rem;
      height: 2rem;
      min-width: auto;
      min-height: auto;
      box-shadow: none;
      border-radius: 50%;
      display: flex;
      justify-content: center;
      align-items: center;
      cursor: pointer;
      margin-left: auto;
    }
  }

  .strikethrough {
    content: "";
    position: absolute;
    display: block;
    width: 100%;
    height: 1px;
    box-shadow: 0 1px rgba(255, 255, 255, 0.6);
    margin-top: 10px;
    background: $green;
    transform-origin: center left;
    animation: strikethrough 1s 0s cubic-bezier(0.55, 0, 0.1, 1) 1;
    transition: transform 0.5s cubic-bezier(0.55, 0, 0.1, 1);
  }
}

@keyframes strikethrough {
  from {
    transform: scaleX(0);
  }
  to {
    transform: scaleX(1);
  }
}
