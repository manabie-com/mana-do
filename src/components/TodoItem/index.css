.ToDo__list {
  display: flex;
  flex-direction: column;
  margin-top: 1.5rem;
}

.ToDo__item {
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-bottom: 10px;
  border-radius: 7px;
  padding: 2px 5px 2px 5px;
}

.ToDo__item input {
  background-color: transparent;
  box-shadow: none;
}

.ToDo__item > span {
  flex: 1 1;
  text-align: left;
  margin-left: 8px;
}
