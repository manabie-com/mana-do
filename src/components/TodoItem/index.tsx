import React from "react";
import { Todo } from "models/todo";
import { isTodoCompleted } from "utils";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTrashAlt } from "@fortawesome/free-solid-svg-icons";

interface TodoItemProps {
  todo: Todo;
  onUpdateStatus: (todoId: string, checked: boolean) => void;
  onDelete: (todoId: string) => void;
}

const TodoItem: React.FC<TodoItemProps> = ({
  todo,
  onUpdateStatus,
  onDelete,
}) => {
  function handleUpdateStatus(
    e: React.ChangeEvent<HTMLInputElement>,
    todoId: string
  ) {
    onUpdateStatus(todoId, e.target.checked);
  }

  return (
    <div className="ToDo__item">
      <input
        type="checkbox"
        checked={isTodoCompleted(todo)}
        onChange={(e) => handleUpdateStatus(e, todo.id)}
      />
      <span>{todo.content}</span>
      <button className="Todo__delete" onClick={() => onDelete(todo.id)}>
        <FontAwesomeIcon icon={faTrashAlt} />
      </button>
    </div>
  );
};

export default TodoItem;
