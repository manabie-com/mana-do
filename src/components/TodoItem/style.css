.ToDo__text{
    padding: 0 10px 0 6px;
    font-weight: 500;
    font-family: Verdana, Geneva, Tahoma, sans-serif;
    cursor: auto;
    overflow: hidden;
    white-space: pre;
    text-overflow: ellipsis;
    max-width: 100%;
}
.ToDo__item{
    height: 25px;
    padding: 10px;
    transition: all 0.4s;
    animation: fly-in 0.2s;
    position: relative;
    word-break: break-all;
    display: flex;
    align-items: start;
    justify-content: space-between;
}
.ToDo__item:hover{
    border-radius: 10px;
    background-color: rgba(245, 245, 245);
    transform: scale(1.02);
    box-shadow: 0px 1px 2px rgba(38, 69, 73, 0.425);;
}
.ToDo__item:hover .ToDo__text_full{
    display: flex;
}

.ToDo__text_full{
    font-weight: 500;
    font-size: 0.9em;
    font-family: Verdana, Geneva, Tahoma, sans-serif;
    display: none;
    background-color: rgb(255, 255, 255);
    padding: 10px;
    border-radius: 10px;
    border: solid 1px rgba(83, 129, 122, 0.589);
    box-shadow: 10px 10px 0 rgba(207, 207, 207, 0.466);;
    max-width: 100%;
    word-break: break-all;
    white-space: normal;
    position: absolute;
    bottom: 110%;
    color: rgb(112, 150, 145);
}

.ToDo__item.active .ToDo__text{
    color: rgb(71, 172, 159);
    text-shadow: 1px 1px 1px rgba(38, 69, 73, 0.425);
}

.ToDo__item.completed .ToDo__text{
    color: rgba(104, 132, 136, 0.753);
    text-decoration: line-through;
}

.ToDo__item.hiding{
    height: 0px;
    padding: 0px;
    opacity: 0;
    animation: fade-out 0.2s;
}

.ToDo__item:last-child::after{
    display: none;
}
.ToDo__item::after{
    content: '';
    position: absolute;
    bottom: 0px;
    background-color: rgba(219, 219, 219, 0.582);
    width: calc(100% - 20px);
    height: 1px;
    opacity: 0.8;
    animation: fade-in-hr 0.6s;
}

.Todo__delete{
    user-select: none;
    outline: none;
    border: none;
    width: 28px;
    height: 28px;
    min-width: auto;
    min-height: auto;
    box-shadow: none;
    border-radius: 50%;
    display: flex;
    justify-content: center;
    align-items: center;
    cursor: pointer;
}
  
  .ToDo__item > span {
    flex: 1 1;
    text-align: left;
    margin-left: 8px;
  }

  .Todo__content {
    flex: 1 1;
  }
  
.Todo__item_input{
    z-index: 10;
    width: 88%;
    min-height: 24px;
    font-size: 1em;
    border: none;
    border-bottom: solid 1px rgb(111, 192, 192);
    color: gray;
    box-shadow: none !important;
    border-radius: 0;
    background: transparent;
    /* background-color: rgba(240, 248, 255, 0.247); */
}

.move-out{
    position: relative;
    animation: fly-out 6s ease-in-out;
}

* {
    font-family: 'Brush Script MT', cursive !important;
}

@keyframes fly-in {
    from{
        height: 0px;
        padding: 0px;
        transform: translate(100px, 60px);
    }
    to{
        height: 25px;
        padding: 10px;
        transform: translate(0px, 0px);
    }
}

@keyframes fly-out {
    0%{
        height: 25px;
        top: 0px;
        left: 0px;
        transform: scale(1);
        padding: 10px;
        opacity: 1;
    }
    5%{
        top: -20px;
        left: 160px;
    }
    10%{
        height: 0px;
        top: 60px;
        left: 200px;
        transform: scale(0);
        padding: 0px;
        opacity: 0;
    }
    100%{
        height: 0px;
        top: 60px;
        left: 200px;
        transform: scale(0);
        padding: 0px;
        opacity: 0;
    }
}

@keyframes fade-out {
    from{
        height: 25px;
        padding: 10px;
        opacity: 1;
        transform: scale(1);
    }
    to{
        height: 0px;
        padding: 0px;
        opacity: 0;
        transform: scale(0.6);
    }
}

@keyframes fade-in-hr {
    from{
        opacity: 0;
        transform: scale(0.4);
    }
    to{
        opacity: 1;
        transform: scale(1);
    }
}