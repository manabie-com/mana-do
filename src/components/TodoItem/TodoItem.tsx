import * as React from "react";
import { useState } from "react";
import {
  KeydownEvent,
  TodoItemProps,
  TodoStatus,
} from "../../models/todo";
import Service from "../../service";
import { setTodos } from "../../store/actions";
import Checkbox from "../Checkbox/Checkbox";
import "./TodoItem.scss";

const TodoItem = ({ todo, dispatch }: TodoItemProps) => {
  const [value, setValue] = useState(todo.content);
  const [isEdit, setEdit] = useState(false);

  const cancelEdit = (
    e:
      | React.FocusEvent<HTMLInputElement, Element>
      | React.KeyboardEvent<HTMLInputElement>
  ) => {
    setValue(todo.content);
    setEdit(false);
    e.preventDefault();
    e.stopPropagation();
  };

  const onUpdateTodoValue = async (
    e: React.KeyboardEvent<HTMLInputElement>
  ) => {
    if (e.key === KeydownEvent.ENTER) {
      todo.content = value;
      const resp = await Service.updateTodo(todo);
      dispatch(setTodos(resp));
      setEdit(false);
    } else if (e.key === KeydownEvent.ESCAPE) {
      cancelEdit(e);
    }
  };

  const onDeleteTodo = async () => {
    const resp = await Service.deleteTodo(todo);
    dispatch(setTodos(resp));
  };

  const checked = todo.status === TodoStatus.COMPLETED;

  const onUpdateTodoStatus = async () => {
    const status = checked ? TodoStatus.ACTIVE : TodoStatus.COMPLETED;
    todo.status = status;
    const resp = await Service.updateTodo(todo);
    dispatch(setTodos(resp));
  };

  return (
    <div
      key={todo.id}
      className="Todo__item"
      onDoubleClick={() => {
        setEdit(true);
      }}
    >
      <div className="Todo__item_checkbox">
        <Checkbox checked={checked} onToggleCheck={onUpdateTodoStatus} />
      </div>
      <div
        className={
          isEdit
            ? "Todo__item_input Todo__item_input--fullwidth"
            : "Todo__item_input"
        }
      >
        {isEdit ? (
          <input
            className="Todo__item_input--modify"
            type="text"
            value={value}
            onChange={(e) => setValue(e.target.value)}
            onBlur={(e) => cancelEdit(e)}
            onKeyDown={(e) => onUpdateTodoValue(e)}
            autoFocus
          />
        ) : (
          <>
            <span className={checked ? "strikethrough" : ""} />
            <span className="Todo__item_input--default">{todo.content}</span>
          </>
        )}
      </div>
      <button className="Todo__item_button--delete" onClick={onDeleteTodo}>
        <img
          src={require("../../assets/images/delete.svg").default}
          alt="deleteIcon"
          height={32}
          width={32}
        />
      </button>
    </div>
  );
};

export default TodoItem;
