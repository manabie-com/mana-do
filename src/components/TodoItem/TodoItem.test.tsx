import React, { FC, ReactElement } from "react";
import { configure, mount } from "enzyme";
import renderer from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";
import { TodoItemProps, TodoStatus } from "../../models/todo";
import TodoItem from "./TodoItem";

configure({ adapter: new Adapter() });

const props: TodoItemProps = {
  todo: {
    content: "1",
    created_date: "2022-04-18T18:18:55.975Z",
    id: "28o6yJPWc3",
    status: TodoStatus.ACTIVE,
    user_id: "firstUser",
  },
  dispatch: jest.fn(),
};

describe("TodoItem", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = mount(<TodoItem todo={props.todo} dispatch={props.dispatch} />);
  });

  afterEach(() => {
    wrapper.unmount();
    jest.clearAllMocks();
  });

  it("Should renders TodoItem correctly", () => {
    const tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should call edit item", () => {
    wrapper
      .find({ className: "Todo__item_input--default" })
      .simulate("doubleClick");
    expect(
      wrapper.find({ className: "Todo__item_input--modify" })
    ).toBeTruthy();
    expect(
      wrapper.find({ className: "Todo__item_input--fullwidth" })
    ).toBeTruthy();
  });
});
