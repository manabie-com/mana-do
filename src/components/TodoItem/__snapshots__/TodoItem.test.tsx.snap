// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem Should renders TodoItem correctly 1`] = `
<div
  className="Todo__item"
  onDoubleClick={[Function]}
>
  <div
    className="Todo__item_checkbox"
  >
    <span
      className="success-checkmark"
      onClick={[Function]}
    >
      <div
        className="icon-circle"
        hidden={false}
      />
      <div
        className="check-icon"
        hidden={true}
      >
        <span
          className="icon-line line-tip"
        />
        <span
          className="icon-line line-long"
        />
        <div
          className="icon-circle"
        />
        <div
          className="icon-fix"
        />
      </div>
    </span>
  </div>
  <div
    className="Todo__item_input"
  >
    <span
      className=""
    />
    <span
      className="Todo__item_input--default"
    >
      1
    </span>
  </div>
  <button
    className="Todo__item_button--delete"
    onClick={[Function]}
  >
    <img
      alt="deleteIcon"
      height={32}
      src="delete.svg"
      width={32}
    />
  </button>
</div>
`;

exports[`TodoItem Should renders TodoItem correctly 2`] = `
<div
  className="Todo__item"
  onDoubleClick={[Function]}
>
  <div
    className="Todo__item_checkbox"
  >
    <span
      className="success-checkmark"
      onClick={[Function]}
    >
      <div
        className="icon-circle"
        hidden={false}
      />
      <div
        className="check-icon"
        hidden={true}
      >
        <span
          className="icon-line line-tip"
        />
        <span
          className="icon-line line-long"
        />
        <div
          className="icon-circle"
        />
        <div
          className="icon-fix"
        />
      </div>
    </span>
  </div>
  <div
    className="Todo__item_input"
  >
    <span
      className=""
    />
    <span
      className="Todo__item_input--default"
    >
      1
    </span>
  </div>
  <button
    className="Todo__item_button--delete"
    onClick={[Function]}
  >
    <img
      alt="deleteIcon"
      height={32}
      src="delete.svg"
      width={32}
    />
  </button>
</div>
`;
